File: mobile/lib/domain/usecases/get_roadmap_use_case.dart
import '../repositories/i_lesson_repository.dart';
import '../repositories/i_learner_repository.dart';
import '../entities/learning_path_node.dart';
import '../personalization/route_engine.dart';

class GetRoadmapUseCase {
  final ILessonRepository lessonRepository;
  final ILearnerRepository learnerRepository;

  const GetRoadmapUseCase(this.lessonRepository, this.learnerRepository);

  Future<LearningPath> execute(String courseId) async {
    final baseNodes = await lessonRepository.getBaseRoadmapNodes(courseId);
    final profile = await learnerRepository.getProfile();
    final scores = profile.scoresByCourse[courseId] ?? {};

    return RouteEngine.adaptRoadmap(
      courseId: courseId,
      baseNodes: baseNodes,
      skillScores: scores,
      completedNodeIds: profile.completedNodeIds,
      completedReinforcementIds: profile.completedReinforcementIds,
    );
  }
}

File: src/domain/usecases/GetRoadmapUseCase.ts
import { ILessonRepository } from '../repositories/ILessonRepository';
import { ILearnerRepository } from '../repositories/ILearnerRepository';
import { LearningPath } from '../entities/LearningPathNode';
import { RouteEngine } from '../personalization/RouteEngine';

export class GetRoadmapUseCase {
  constructor(
    private lessonRepo: ILessonRepository,
    private learnerRepo: ILearnerRepository
  ) {}

  async execute(courseId: string): Promise<LearningPath> {
    const baseNodes = await this.lessonRepo.getBaseRoadmapNodes(courseId);
    const profile = await this.learnerRepo.getProfile();
    const scores = profile.scoresByCourse[courseId] || {};

    return RouteEngine.adaptRoadmap(
      courseId,
      baseNodes,
      scores,
      profile.completedNodeIds || [],
      profile.completedReinforcementIds || []
    );
  }
}
